common/persistence/history_manager.go

```diff
@@ -361,6 +372,35 @@ func (m *executionManagerImpl) getHistoryBranchInfo(
 	return resp.BranchInfo, err
 }
 
+func ToHistoryBranchInfo(serializer serialization.Serializer, blob *commonpb.DataBlob) (*persistencespb.HistoryTreeInfo, error) {
+	treeInfo, err := serializer.HistoryTreeInfoFromBlob(blob)
+	if err != nil {
+		return nil, err
+	}
+	// For existing data that was previously persisted with only branch info but without the opaque branch token,
+	// compute the missing branch token on the fly. Similarly, if only opaque branch token exists but without the branch
+	// info, then compute the missing branch info on the fly.
+	// TODO: Consider removing in the future, but this will be a forward/backward incompatible change for old data.
+	if treeInfo.BranchToken == nil {
+		blob, err := serializer.HistoryBranchToBlob(treeInfo.BranchInfo, enumspb.ENCODING_TYPE_PROTO3)
+		if err != nil {
+			return nil, err
+		}
+		treeInfo.BranchToken = blob.Data
+	} else if treeInfo.BranchInfo == nil {
```

Comment:
Why we need fill in the `BranchInfo`? I thought all new logic is going to use BranchToken field? and we do double write to be compatible with the old code.

Reply:
I tried to describe this in the git message body. Basically, if some customer were to upgrade a live deployment from before this change to after this change, then they wouldn't be able to downgrade unless we double write. At some point (in the medium~distant future) this double write will be removed and customers won't be able to downgrade to before this change.